ctlcmd/cmdbase.py

```diff
@@ -771,7 +773,189 @@ def globcomp(text):
+
+    # Adding time stamp filenames
+    timestring = datetime.datetime.now().strftime('%Y%m%d_%H%M%S')
+    filename = re.sub('<TIMESTAMP>', timestring, filename, flags=re.IGNORECASE)
+
+    # Adding boardid to the settings
+    filename = re.sub('<BOARDID>',
+                      str(self.board.boardid),
+                      filename,
+                      flags=re.IGNORECASE)
+    # Adding boardtype to the settings
+    filename = re.sub('<BOARDTYPE>',
+                      str(self.board.boardtype),
+                      filename,
+                      flags=re.IGNORECASE)
 
+    # Substituting tokens for argument values
+    for action in self.parser._actions:
+      for string in action.option_strings:
+        string = string.strip('-')
+        if not hasattr(args, string): continue
+        val = ''
+        try:
+          val = action.type(getattr(args, string))
+        except:
+          pass
+        substring = '{0}{1}'.format(string, val)
+        substring = re.sub(r'\.', 'p', substring)
+        filename = re.sub('\\<{0}\\>'.format(string),
+                          substring,
+                          filename,
+                          flags=re.IGNORECASE)
+    self.saveroot = filename
+    return args
+      
+  def maketree(self,data,datatypes={}):
+    """
+    @brief Create the root file and the TTree in the root file
+
+    @details This command creates the root file and creates the TTree with the correct
+    data types and names for each branch of the tree. This also saves some extra basic data 
+    like the time the file was created, the Board_ID and Board_Type. This function is only called once.
+
+    """
+     
+    self.rootfile = uproot.recreate(self.saveroot)
+    standard_dict={"time":np.float32,"det_id":int,"gantry x":np.float32,"gantry y":np.float32,
+                   "gantry z":np.float32,"LED bias voltage":np.float32,"LED temp":np.float32,"SiPM temp":np.float32} 
+    self.specific_dict={**{title:type(data[title]) for title in data}}
+
+    ##manual override if type is not returning a value that root will accept
+    if datatypes:
+      for title in datatypes:
+        self.specific_dict[title]=datatypes[title]
+    
+    self.rootfile.mktree("DataTree",{**{title:standard_dict[title] for title in standard_dict},**{title:self.specific_dict[title] for title in self.specific_dict}})
+
+    ##add extra data which must only be saved once
+    timestring = datetime.datetime.now().strftime('%Y%m%d_%H%M%S')
+    self.rootfile["Board_ID"]=self.board.boardid
+    self.rootfile["Board_Type"]=self.board.boardtype
+    self.rootfile["Time"]=timestring
+    
+    ##Create apparatus to store data until it is pushed to the root file
+    self.n=0
+    standarddata={**{title:[] for title in standard_dict}}
+    specificdata={**{title:[] for title in self.specific_dict}}
+    self.saveddata={"standard":standarddata,"specific":specificdata} 
+  
+  def fillroot(self,data,datatypes={},time=0.0,det_id=-100):
```

Comment:
It is dangerous to have mutable data types as the default argyment in python [1]

You should change it to `datatypes=None` then add a `datatypes = {} if datatypes is None else datatypes` as the first statement of the function.

[1] https://stackoverflow.com/questions/1132941/least-astonishment-and-the-mutable-default-argument